File: Data Structures & Algorithms/combinations/submission-3.py
class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        res = []

        def generateCombinations(i, curr):
            if len(curr) == k:
                res.append(curr[:])
                return

            if len(curr) + (n - i + 1) < k:
                return

            curr.append(i)
            generateCombinations(i + 1, curr)

            curr.pop()
            generateCombinations(i + 1, curr)

        generateCombinations(1, [])
        
        return res

File: Data Structures & Algorithms/permutations/submission-3.ts
class Solution {
    /**
     * @param {number[]} nums
     * @return {number[][]}
     */
    permute(nums: number[]): number[][] {
        let res: number[][] = [[]]

        for (const n of nums) {
            const curr = []
            for (const perm of res) {
                for (let i = 0; i <= perm.length; i++) {
                    const next = [...perm]
                    next.splice(i, 0, n)
                    curr.push(next)
                }
            }
            res = [...curr]
        }
        return res
    }
}
